Updated AMD benchmarks

MikhailRyazanov opened the pull request:
I've got a new computer, with AMD Ryzen 5 5600G, and ran all the speed benchmarks on it. I guess, it would be better use them to replace the old Ryzen 3 2200G results (particularly because the 2200G data lacks `direct_C` and `daun`, and they can't be added now, since the CPU in that machine has been changed to 3200G).

The plotting code is also slightly simplified, plus `pyabel-timing-extra.py` is added for “non-systematic” benchmarks (currently only for running `daun` with `reg='nonneg'` on the O<sub>2</sub><sup>−</sup> example).

MikhailRyazanov commented:
Unexpectedly, RTD has caused too many problems. First, an unannounced requirement to specify the build environment, then the fact that they're still pinning our “old” project to [very old package versions](https://docs.readthedocs.io/en/stable/build-default-versions.html), which at this point became incompatible with everything else... So I had to update the required Sphinx and RTD theme versions to make it working at all (this also brings them closer to local builds). Hopefully, this didn't break the resulting documentation, but please take a look at everything you can (I've looked quickly at some sections and didn't see anything obviously wrong, but I'll also try to check thoroughly).

stggh commented:
@MikhailRyazanov it is great that you have a new toy. Thanks for delving into all the subsequent issues, your solutions are very informative and useful.

I am running a few tests, and will get back to this discussion later.

DanHickstein commented:
Awesome! Our best benchmarks yet :). Now all of the "real" methods (which excludes Direct and Bordas) are solidly capable of HD video rates. I looked at the new docs and they don't seem obviously broken. I'd say we can just merge and deal with any small bugs with the docs later.

MikhailRyazanov commented:
Dan, thanks for checking this! In the meantime, I ran `linkcheck`, which has detected that some external links have moved/disappeared, so I've updated them correspondingly. I've also accidentally noticed some typos and corrected them. @stggh, could you please confirm that `O-ANU1024` should really be `O2-ANU1024`?

As for the “best benchmarks”, it seems that many but _not all_ timings are better that those from i7-9700, and my guess is that some are mostly limited by the RAM speed (these got better), but some depend mostly on the CPU and code optimizations, so they are still faster on i7-9700 with MKL.

stggh commented on abel/tools/vmi.py:
```diff
@@ -441,7 +441,7 @@ def toPES(radial, intensity, energy_cal_factor, per_energy_scaling=True,
         eV/pixel\ :sup:`2`, will give output energy units in eV.  A value of
         :math:`1.148427\times 10^{-5}` cm\ :math:`^{-1}/`\ pixel\ :sup:`2`
-        applies for "examples/data/O-ANU1024.txt" (with Vrep = −98 volts).
+        applies for "examples/data/O2-ANU1024.txt" (with Vrep = −98 volts).
```
No, the discussion and  scaling factor actually apply to O<sup>-</sup> photodetachment, while, in the past, the data for `example_hansenlaw.py` was changed to `O2-ANU1024.txt`, no doubt to reduce the PyAbel package size. I will update the `abel.tools.vmi.toPES` text during today.

MikhailRyazanov replied:
OK, now I see that it was indeed `O-` but then was replaced by `O2-` here: ff35dbf. Perhaps then all the references to it should be adjusted accordingly, or maybe `SampleImage(n=..., name='Ominus')` can be used instead?

On the other hand, we actually don't include any example data in the package, so it shouldn't be a problem to readd that file back, if it's useful at all.

One strange thing that I've just noticed is that the PyPI package does include the `examples` directory, although without `examples/data` and even without `examples/README.txt`, but the conda packages don't include any examples (which is probably expected). @DanHickstein, what do you think we should do about this? I guess, providing `pyabel-examples` as a separate package (with the data files) might be useful, but I have no idea how this can be done easily...

DanHickstein replied:
Thanks for catching the typos!

I think that it's okay if we don't install the examples in the case of PyPi or Conda install, since most users will be interacting with the examples as part of the docs. I'm actually a bit surprised that they are installed with the PyPi install, since I think that we previously observed that they were not installed:
https://github.com/PyAbel/PyAbel/issues/26#issuecomment-170242549
But I don't have strong feelings on the issue.

What is the motivation for moving the examples to a separate repo? I think that the standard practice is to keep them in the same repository. For example, matplotlib has them in the main repo: https://github.com/matplotlib/matplotlib/tree/main/galleries

stggh replied:
Here is a diff for `abel/tools/vmi.py` with the corrected constants for O<sub>2</sub><sup>-</sup>:

```
diff vmi.py ~/Downloads/.
422,424c422,424
<     generated energy scale with published spectra. e.g. for O\ :sub:`2`\ :sup:`−`
<     photodetachment, the origin band occurs at the
<     electron affinity :math:`EA = 3,613` cm\ :math:`^{-1}`. Values for
---
>     generated energy scale with published spectra. e.g. for O\ :sup:`−`
>     photodetachment, the strongest fine-structure transition occurs at the
>     electron affinity :math:`EA = 11\,784.676(7)` cm\ :math:`^{-1}`. Values for
443,444c443,444
<         :math:`1.204\times 10^{-5}` cm\ :math:`^{-1}/`\ pixel\ :sup:`2`
<         applies for "examples/data/O2-ANU1024.txt" (with Vrep = −2200 volts).
---
>         :math:`1.148427\times 10^{-5}` cm\ :math:`^{-1}/`\ pixel\ :sup:`2`
>         applies for "examples/data/O2-ANU1024.txt" (with Vrep = −98 volts).
458c458
<         is :math:`10^{7}/\lambda` (nm) e.g. ``1.0e7/454.5`` for
---
>         is :math:`10^{7}/\lambda` (nm) e.g. ``1.0e7/812.51`` for
466c466
<         is applied. e.g. ``-2200`` (volts), for "examples/data/O2-ANU1024.txt".
---
>         is applied. e.g. ``-98`` (volts), for "examples/data/O2-ANU1024.txt".
```

MikhailRyazanov replied:
Thanks, @stggh! I've incorporated your corrections (with minor formatting changes, to avoid line breaks between “cm” and “⁻¹”), please check that now they all are as expected.

stggh commented:
A few things have changed over time:
- [daz](https://github.com/chainer/daz) no longer installs, due to a missing `imp` python module.
- [MKL](https://anaconda.org/anaconda/mkl) "should" be included within Anaonda Python, but doesn't appear in `.show_config()`,

and I just had to re-run some of the timings. For some reason the i7-6700 timings have improved (not listed here). 
Any timing updates may be included later. 

The new Ryzen5 starts off faster, but then converges to the i7-9700 result. The OrangePi5 does well in the bang/$ category.

![cf](https://github.com/PyAbel/PyAbel/assets/10932229/802f3b09-47c4-4f0d-b64b-830ddfe4f1a1)
![cf](https://github.com/PyAbel/PyAbel/assets/10932229/5343004e-53d8-4469-b149-670aac033a2f)

```
(py312)[Gibbo:.+/fig_benchmarks]$ cat working/hansenlaw.dat 
# hansenlaw       iabel(ms)           
    5             0.10410
    9             0.12103
   17             0.16169
   33             0.24174
   65             0.43798
  129             0.99764
  257             2.79352
  513             9.03836
 1025            33.32467
 2049           121.10773
 4097           510.34561
 8193          1574.43416
16385          5987.98083
32769         36042.93715

(py312)[Gibbo:.+/fig_benchmarks]$ cat Ryzen5-5600G_Linux/hansenlaw.dat 
# hansenlaw       iabel(ms)           
    5             0.09188
    9             0.10680
   17             0.14250
   33             0.21161
   65             0.37621
  129             0.75968
  257             1.80213
  513             5.06276
 1025            19.95768
 2049            76.44536
 4097           300.99862
 8193          1443.36033
16385          8271.27610
32769         31234.13253

(py312)[Gibbo:.+/fig_benchmarks]$ cat OrangePi5_armbian/hansenlaw.dat 
# hansenlaw       iabel(ms)           
    5             0.21973
    9             0.25346
   17             0.33597
   33             0.49767
   65             0.89308
  129             1.82993
  257             4.82953
  513            15.61482
 1025            59.63636
 2049           309.54884
 4097          1379.63771
 8193          6892.70024
16385         35458.90235
```

stggh commented on .readthedocs.yml:
```diff
@@ -5,6 +5,11 @@
 # Required
 version: 2
 
+build:
```
Good to know this stuff. Thanks

stggh commented on doc/transform_methods/comparison/fig_benchmarks/pyabel-timing-extra.py:
```diff
@@ -0,0 +1,38 @@
+# The following should be used for Intel CPUs, since they work very slowly
+# with subnormal floating-point numbers (see PyAbel issue #246):
+try:
+    import daz
```
Current [daz](https://github.com/chainer/daz) does not install

stggh reviewed (approved):
Good work @MikhailRyazanov. This PyAbel section is difficult to maintain, due to the many system variables: `daz`, `MKL`, and python developments in each new version.

MikhailRyazanov commented:
@stggh, thanks for noticing the problem with `daz`! I was using it recently without installing (it did compile and work). Unfortunately, the original repository doesn't seem to be maintained any more, so I'll need to investigate whether there are now any reliable alternatives (maybe even in NumPy or BLAS)...

Let's merge this now, then see what else can be done...
Do you think that adding Orange Pi benchmarks would be worthy?

stggh commented:
> Do you think that adding Orange Pi benchmarks would be worthy?

After my raspberry pi 5 arrives, in a few weeks, we will see.

Merge for now.
